fix basic auth; fix listing lists; error handling

diablodale opened the pull request:
**fix basic auth**
api index and therefore the routes wouldn't load because code didn't pass auth as needed at that stage

**fix lists**
when using the command "list" the code was not handling some lists being returned from WP 4.7.2. For example:
- wp-cli --http=https://user:password@server.domain.com rest settings list 
- wp-cli --http=https://user:password@server.domain.com rest status list

**error handling of outputs**
Added error handling of outputs. E.g., if an error occurs, then --porcelain does not output data.

Dev tested against WP 4.7.2 on Ubuntu 16.04 with PHP 7.0.13

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -61,9 +61,15 @@ public function set_auth( $auth ) {
 		list( $status, $body ) = $this->do_request( 'POST', $this->get_base_route(), $assoc_args );
 		if ( Utils\get_flag_value( $assoc_args, 'porcelain' ) ) {
-			WP_CLI::line( $body['id'] );
+			if ( $status < 400 ) {
```
Can you add a test for this change?

It'd probably be good to explicitly check for `$body['id']` being set too

diablodale replied:
Caveat, I am learning Travis on the fly. Might see several commits as I figure out how to use it.
I found examples where the REST api returns list results with no "id". I wasn't able to reverse-out the spec on what to do in those situations. So in this case, what *is* the spec/rule? Here's the scenario
- user types command line to create an item
- code calls a REST api to create item. code expects an id to be returned
- code receives an http result code < 400 yet no id
What is the output for the case with and without porcelain?

It is possible for REST to return http=200, a body, and no id. So that means the server did...something. I intentionally left the code in a manner that errors will occur so that the admin can see something unexpected happened. Think of porcelain, in that usage I would output nothing. A user then knows the create failed yet has no diagnostics yet the server did something. I didn't like that scenario so I was allowing the code to fault so it could provide diagnostics.
But...what do you think. What is the approach/spec on this? For both porcelain and without?

diablodale replied:
Would you please add apache to the travis config? It would likely much churn for me to get that working. The docs to do it are at https://docs.travis-ci.com/user/languages/php/#Apache-%2B-PHP
Why? None of the test cases use the --http code path. They are all calling locally. The code paths in restful are significantly different between them. And the results coming back from REST are sometimes different. I encountered this during my testing and handled the differences in code.
With apache in place, I can add test cases which exercise the new PRs I'm making.

danielbachhuber replied:
> Caveat, I am learning Travis on the fly. Might see several commits as I figure out how to use it.

You should be running the tests locally. Here's the appropriate documentation: https://make.wordpress.org/cli/handbook/pull-requests/

> But...what do you think. What is the approach/spec on this? For both porcelain and without?

Porcelain is meant to only ever output the result of the operation that can be piped to another command. I think we should continue with this behavior. If RESTful WP-CLI sometimes doesn't return output, the exit code should indicate whether it was an error or not.

> Would you please add apache to the travis config? It would likely much churn for me to get that working.

You're welcome to work on it in a separate pull request. I don't have the bandwidth to handle this immediately.

danielbachhuber replied:
Related #10

diablodale replied:
error codes now set in exit code and/or error text. Put error code at end of line without trailing period to make it easier for parsing.

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -61,9 +61,15 @@ public function set_auth( $auth ) {
 		} else {
-			WP_CLI::success( "Created {$this->name} {$body['id']}." );
+			if ( $status < 400 ) {
+				WP_CLI::success( "Created {$this->name} {$body['id']}." );
```
Let's also check that `$body['id']` is set before attempting to display it.

diablodale replied:
done

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -61,9 +61,15 @@ public function set_auth( $auth ) {
+			if ( $status < 400 ) {
+				WP_CLI::success( "Created {$this->name} {$body['id']}." );
+			} else {
+				WP_CLI::error( 'Could not complete request.' );
```
Let's include the HTTP code in this error message:

```
WP_CLI::error( "Could not complete request (HTTP code {$status})" );
```

diablodale replied:
error codes now set in exit code and/or error text. Put error code at end of line without trailing period to make it easier for parsing.

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -80,25 +86,32 @@ public function generate_items( $args, $assoc_args ) {
+				if ( Utils\get_flag_value( $assoc_args, 'porcelain' ) ) {
+					return;
 				}
+				WP_CLI::error( 'Could not complete request.' );
```
It'd be good to include the http code here too (similar to above).

diablodale replied:
error codes now set in exit code and/or error text. Put error code at end of line without trailing period to make it easier for parsing.

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -111,12 +124,19 @@ public function generate_items( $args, $assoc_args ) {
+				}
 			} else {
-				WP_CLI::success( "Deleted {$this->name} {$body['id']}." );
+				WP_CLI::error( 'Could not complete request.' );
```
Same as above.

diablodale replied:
error codes now set in exit code and/or error text. Put error code at end of line without trailing period to make it easier for parsing.

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -145,12 +165,30 @@ public function get_item( $args, $assoc_args ) {
 				'api_url'     => $this->api_url,
 			) );
 		} else {
+			if ( empty( $body ) ) return;
```
Coding standards: `return` should be on a new line, and surrounded by braces.

diablodale replied:
ok

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -292,9 +336,15 @@ public function diff_items( $args, $assoc_args ) {
+			if ( $status < 400 ) {
+				WP_CLI::success( "Updated {$this->name} {$body['id']}." );
+			} else {
+				WP_CLI::error( 'Could not complete request.' );
```
Same as above.

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -428,7 +478,7 @@ private function do_request( $method, $route, $assoc_args ) {
 			$body = json_decode( $response->body, true );
 			if ( $response->status_code >= 400 ) {
 				if ( ! empty( $body['message'] ) ) {
-					WP_CLI::error( $body['message'] . ' ' . json_encode( array( 'status' => $response->status_code ) ) );
```
Why'd you remove this?

diablodale replied:
I didn't. It is still there. There was only a change to no longer put the status at the end so that it matched the output of your test cases.
If you want to change the spec on output, then the code and the test cases need to change. I would recommend that a consistent manner of reporting http status be chosen. In the old codebase, it literally shows the JSON. In your PR feedback above, you want it be not show JSON and be formatted. That would cause inconsistencies in http error code output which I would discourage.

danielbachhuber commented on inc/Runner.php:
```diff
@@ -85,13 +88,9 @@ public static function after_wp_load() {
-			return false;
-		}
+		if ( $response->status_code >= 400 ) return false;
+		if ( empty( $response->headers['link'] ) ) return false;
```
Coding standards

diablodale replied:
ok

danielbachhuber commented on inc/Runner.php:
```diff
@@ -105,13 +104,16 @@ private static function auto_discover_api( $url ) {
 		}
+		$response = Utils\http_request( 'GET', $api_url, null, $headers );
+		if ( $response->status_code >= 400 ) return false;
+		if ( empty( $response->body ) ) return false;
```
Coding standards

diablodale replied:
ok

danielbachhuber commented:
@diablodale Out of curiosity, how are you using this package? Testing it out, or for a project in the real world? And, what do you think about its utility generally?

diablodale commented:
I'm using wp-cli in AWS. Specifically, in a CloudFormation template used to create my website/eStore infrastructure. I was using just wp-cli but found a scenario where I couldn't do the work "local" due to process access/ownership of my cache. I needed to call to the webserver so that normal handling of input would correctly update the cache. When I saw the restful package, I started testing its usage and found many list scenarios were not working. Which led to this PR.
With WP core committed to REST, I do think a command line tool to manipulate that API is needed. There is/will be people that want to automate WP activities at the command-line. I do not know the number of people.

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -188,8 +254,14 @@ public function list_items( $args, $assoc_args ) {
 				'api_url'     => $this->api_url,
 			) );
 		} else {
+			if ( empty( $items ) ) return;
```
This should be:

```
if ( empty( $items ) ) {
	return;
}
```

diablodale replied:
ok

danielbachhuber commented on features/comment.feature:
```diff
@@ -2,7 +2,8 @@ Feature: Manage WordPress comments through the REST API
     Given a WP install
-    And I run `wp plugin install rest-api --activate`
+    When I run `wp core version`
+	Then STDOUT should be a version string >= 4.7
```
Minor nit: it looks like your indentation is off here.

Bigger question: Why is testing the version important?

diablodale replied:
Because the REST api is not part of the core WP codebase until version 4.7. I want to ensure that we have WP 4.7 or greater. These test cases will fail otherwise.

danielbachhuber replied:
> Because the REST api is not part of the core WP codebase until version 4.7. I want to ensure that we have WP 4.7 or greater. 

I think this is an assumption we don't need to test for. The test suite downloads the latest version of WordPress behind the scenes. We can assume we're on the latest version of WordPress (and that we have a working database, etc.).

diablodale replied:
I disagree. The test cases are separate from travis. And separate from WP installs. The cases a behat and run separately (like I've done forever since I started using these test tools).
Behat docs suggest that the background givens ensure that the environment is setup as expected. It is not guaranteed (since these behat cases are separate) that WP is any specific version. Therefore, the test for the specific version.
Since this only makes the test cases stronger, I can't understand why you would not allow this to remain. It is not enough to just have a WP install. It must be a WP 4.7+ install.

danielbachhuber replied:
> Since this only makes the test cases stronger, I can't understand why you would not allow this to remain. It is not enough to just have a WP install. It must be a WP 4.7+ install.

This specific comment isn't a strongly held opinion. However, I must point you to my previous comment on this pull request:

> At this point, I'm not confident in accepting the entire pull request. Some changes make sense, and others need more consideration. I'd like for it to be broken into logical subparts, so that each piece can be evaluated on its own.

diablodale replied:
Of course. The pull request is for your benefit and those that use restful at large. I will continue to use my restful fork until some future where you reach my functionality level or we merge. I have a well working fork and my needs as such that I won't wait or lessen functionality. That's one of the benefits of forks. :-)

danielbachhuber replied:
Ok, that's fine. I'll pull code out of your pull request when I have some bandwidth for it.

danielbachhuber commented on features/comment.feature:
```diff
@@ -152,8 +154,9 @@ Feature: Manage WordPress comments through the REST API
-      Error: Sorry, you can not delete this comment
+      Error: Sorry, you are not allowed to delete this comment. HTTP code: 401
       """
+    Then the return code should be 145
```
I'm still not sure how I feel about using these elevated return codes (WP-CLI still uses `0` vs `1`). It doesn't need to block the PR though; we can discuss separately #96

diablodale replied:
comments in #96

danielbachhuber commented on features/load-rest.feature:
```diff
@@ -2,7 +2,8 @@ Feature: Manage WordPress through endpoints locally
     Given a WP install
-    And I run `wp plugin install rest-api --activate`
+	When I run `wp core version`
+	Then STDOUT should be a version string >= 4.7
```
Indentation.

diablodale replied:
I see no issue all tests pass. Perhaps this is a text/tab oddity

danielbachhuber replied:
> I see no issue all tests pass. Perhaps this is a text/tab oddity

Behat `.feature` files require spaces, not tabs. We don't have a linter in place right now.

diablodale replied:
behat doesn't require spaces. It only requires that related indentation be the same number of chars. If you are suggest this is a coding-style rule, then the coding style rules need to be updated to indicate that tabs are disallowed in xxx, xxx, xxx set of files. The php rules at https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation declare tabs should be used. Since this is the primary language used for wp-cli developers, it seems odd to me to change the indentation rule which then creates an additional burden on devs and reviewers both. Consistency helps efficiency and decreases error rates.

danielbachhuber replied:
> If you are suggest this is a coding-style rule, then the coding style rules need to be updated to indicate that tabs are disallowed in xxx, xxx, xxx set of files.

See https://github.com/wp-cli/restful/blob/master/.editorconfig#L16

> The php rules at https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation declare tabs should be used.

`.feature` != `.php`

danielbachhuber replied:
I've filed an issue to include a linter in the future https://github.com/wp-cli/scaffold-package-command/issues/82

diablodale replied:
sounds good. tools and automation to help devs is usually a great idea. Uuusually... ;-)

diablodale replied:
A caution: this .editorconfig file is useless to me. It is also a hidden requirement and functionality as it is not documented in the coding standards. My editors (e.g. VS Code, notepad++, and vi) do not use such a file. On searching now, it is possible to install a 3rd party extension in VS Code for it to work. However, if that is needed, then such needs to be documented in the coding standards.

danielbachhuber commented on features/steps/given.php:
```diff
@@ -102,7 +102,7 @@ function ( $world, TableNode $table ) {
 );
 
-$steps->Given( '/^save (STDOUT|STDERR) ([\'].+[^\'])?as \{(\w+)\}$/',
+$steps->Given( '/^save (STDOUT|STDERR) ([\'].+[^\'])?\s?as \{(\w+)\}$/',
```
This source of this code is WP-CLI's test suite. Want to submit a PR there too? https://github.com/wp-cli/wp-cli/blob/master/features/steps/given.php#L112

Related https://github.com/wp-cli/wp-cli/issues/3826

diablodale replied:
done

danielbachhuber commented on features/steps/then.php:
```diff
@@ -145,6 +145,16 @@ function ( $world, $stream ) {
 	}
 );
 
+$steps->Then( '/^(STDOUT|STDERR) should be a version string (<|<=|>|>=|==|=|!=|<>) ([+\w\.-]+)$/',
```
This will eventually want to end up in WP-CLI's test suite, if we deem it a necessary test to keep.

diablodale replied:
issue opened, fixed, and soon PR

danielbachhuber commented on inc/RestCommand.php:
```diff
@@ -111,12 +133,43 @@ public function generate_items( $args, $assoc_args ) {
+	 *
+	 * @subcommand delete
+	 */
+	public function purgeall_items( $args, $assoc_args ) {
```
Did you mean to include this?

diablodale replied:
Yes, the text comment should read purgeall instead. I have updated it.

danielbachhuber commented:
@diablodale First of all, I sincerely appreciate the effort you've put into this pull request. Please don't take my comments as disrespect to the time you've spent.

It's getting harder to approve the pull request because of the amount of code that's changed. While this isn't necessarily a bad thing at this stage (the package probably isn't in production use many places), I'm concerned about inheriting decisions, bugs, etc. that we have to live with indefinitely.

To make it easier to approve the changes going in, would you be willing to separate the pull request into logical subparts? We could start with the changes required to get the tests passing again, then basic auth, and then the new features you want to add separately.

diablodale commented:
All tests pass as is. No failures.
I recommend taking the current set of commits as a whole. They have bug fixes throughout as I made needed corrections to the scaffolding. What you and others wrote was good; it provided a base from which I could get it working. As I continued my testing, I made the fixes.

The "contract" is the output. I have not taken away features. Instead, I fixed existing ones and enabled intended ones (like 2+ path param endpoints). So...by looking at the changes to the test cases, you can see where things would change for users on the internet. Other things I implemented they couldn't be using...because they didn't work before this set of commits.

click over on the tab and you can see what changed in the 3 test files. In short, it is a change in the capitalization of "x-wp-totalpages" and a change to the error text. Everything else is the same.

danielbachhuber commented:
> The "contract" is the output. I have not taken away features. Instead, I fixed existing ones and enabled intended ones (like 2+ path param endpoints).

I understand this. However, the policy is to [open a new issue to discuss features](https://make.wordpress.org/cli/handbook/contributing/#creating-a-pull-request) and, generally, to keep to one feature per pull request.

At this point, I'm not confident in accepting the entire pull request. Some changes make sense, and others need more consideration. I'd like for it to be broken into logical subparts, so that each piece can be evaluated on its own.

diablodale commented:
Oh, and there are bugs still in the codebase areas I touched. None that affect the functionality that I'm using and the basic functionality that I enabled. Some need discussion, like colliding param names. Some are cosmetic and will live to another day when someone has time to address them.

If you find bugs which affect the direct functionality that I fixed or enabled, then YES I want to isolate and fix them. That's my responsibility as I'm the source of them.

I am now writing the custom endpoint which is then exposed with the restful functionality that I enabled. It is a good breaking point, a way to segment that work (which I've already started) from the work in this PR and batch of commits.

diablodale commented:
Of course your choice as it is your repro. I'll continue in my fork and when you want to collaborate, please loop me in.
